### PR TITLE
Fix minor things on the Faas + Astra DB/Cassandra blog post (2021-08-26)

### PR DESCRIPTION
**Pull request by hemidactylus**
- remark on "create keyspace" command
- insert statements, missing punctuation
- redundant (kubectl) secret generation
- out-of-order dir navigation command
- env var key for providing keyspace name to function
- document-API GET call would fail with a 404

Signed-off-by: Stefano Lottini <stefano.lottini@datastax.com>

## Description
Six changes to the blog post about OpenFaas+AstraDB/Cassandra to ensure commands succeed for those who follow along and execute the instructions in the blog line by line:

* CREATE KEYSPACE cql commands work in vanilla Cassandra, but in Astra you create keyspaces through the UI (added a remark in that sense)

* Cyclist insert statements: missing commas and mismatched column/values were giving syntax error

* Secrets: removed a "kubectl create secret" command, not needed and was duplicating a secret created with faas-cli, hence gave an error

* deploying the cycling.yml: "cd .." should come right before and not right after (for consistence with previous console navigation)

* weekly-newsletter.yml, correct key is ASTRA_DB_KEYSPACE instead of ASTRA_DB_APPLICATION_TOKEN

* Removed the first ("Check the table is empty:") part since that would 404 until a document is first inserted, as which point the underlying collection is actually created (I checked with the eng team, and it's by design)

## Motivation and Context
I reproduced the process described in the blog, ran the commands given there one by one and noticed some minor flaws that, once improved, would have increased the quality of the post.

## Have you applied the editorial and style guide to your post?
Just minor editing to an existing post, so: yes.

## How have you tested the instructions for any tutorial steps?
Reproduced the whole experience reported in the post, twice, taking notes, with:
- a KinD cluster (v0.11.1 go1.16.6 linux/amd64)
- OpenFaas (CLI 0.13.13, Gateway 0.21.1, Provider 0.13.5)
- an Astra DB instance
- (Ubuntu Linux 20.04 64-bit)

## Types of changes

- [ ] New blog post
- [X] Updating an existing blog post
- [ ] Updating part of the page page
- [ ] Adding a new web-page

## Checklist:

- [X] I have given attribution for any images I have used and have permission to use them under Copyright law
- [X] My code follows the [writing-style of the publication](README.md) and I have checked this
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`

### CODE REVIEW

File: _posts/2021-08-26-faas-storage-cassandra-astra.md [hunk 1]
```diff
@@ -89,23 +91,23 @@ You can switch keyspace with the `use KEYSPACE;` command such as `use cycling;`
 ```sql
 // Insert a cyclist record record
 INSERT INTO cycling.cyclist_name (
-  id
-  lastname
+  id,
```
Comment (alexellis) [inline _posts/2021-08-26-faas-storage-cassandra-astra.md:94]: This would have been copied and pasted, so why did it work for me, but not for you?

File: _posts/2021-08-26-faas-storage-cassandra-astra.md [hunk 2]
```diff
@@ -434,15 +433,6 @@ There's three ways to use the function:
 
 In production, you'll also want to add authentication to the submission endpoint. You can learn how with my Serverless For Everyone Else eBook listed at the end of the article. Here, I just want to focus on getting you connected and getting/putting documents into Astra DB.
 
-Check the table is empty:
```
Comment (alexellis) [inline _posts/2021-08-26-faas-storage-cassandra-astra.md:437]: Why was this removed?

File: _posts/2021-08-26-faas-storage-cassandra-astra.md [hunk 3]
```diff
@@ -179,9 +181,6 @@ faas-cli secret create astra-clientsecret \
 faas-cli secret create astra-secure-connect \
 --from-file astra-secure-connect
 
-kubectl create secret generic -n openfaas-fn astra-secure-connect \
```
Comment (alexellis) [inline _posts/2021-08-26-faas-storage-cassandra-astra.md:182]: This can be removed, I ran into an issue with openfaas that was patched as part of this blog post. faas-cli secret create will work with the latest few versions now :+1:

File: _posts/2021-08-26-faas-storage-cassandra-astra.md (hunk 1 above)
Comment (hemidactylus) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:94]: All I can offer is that perhaps some commas were lost in the final editing of the post. CQL syntax requires the column list to be comma-separated (https://docs.datastax.com/en/cql-oss/3.3/cql/cql_reference/cqlInsert.html), indeed if I paste the statement as it is (both on standard Cassandra and Astra CQL Console) I get the syntax error
`SyntaxException: line 3:2 mismatched input 'lastname' expecting ')' (...INTO cycling.cyclist_name (  id  [lastname]...)`

Same for the fact that the values list has to be in the same order as the column list provided earlier: if the orderings don't correspond, CQL complains with something like
`InvalidRequest: Error from server: code=2200 [Invalid query] message="Type error: cannot assign result of function system.uuid (type uuid) to firstname (type text)"`

(I just tried with a regular Cassandra cluster as well as Astra DB to make sure both reproduce the same syntax errors consistently, and they do.)

File: _posts/2021-08-26-faas-storage-cassandra-astra.md (hunk 2 above)
Comment (hemidactylus) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:437]: The reason is that, in the Document API by Stargate (broadly modeled after other existing document databases), if the first action on a not-yet-existing collection is a write, the collection is automatically created for you; conversely, if the first action is a read, it fails with a 404 error.
This is a by-design behaviour of the underlying Stargate gateway which the `@astrajs/collections` SDK inherits: so, if the collection does not exist yet, this first read would fail with a 404.

I have to add that I did not know that beforehand, and had to inquiry with the engineering team to find out the rationale behind this choice; also, it is easy to overlook the not-found error on first pass (as happened to me on the first run) since it does not hinder anything of what follows (indeed the collection is created in the next step and everything proceeds flawlessy since then).

Comment (alexellis) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:437]: Thanks for the explanation.

That is something that we should fix in the function then.. do you want to try revising the Node.js code?

The API (function) shouldn't leak the implementation of the document database underneath.

File: _posts/2021-08-26-faas-storage-cassandra-astra.md (hunk 1 above)
Comment (alexellis) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:94]: Thanks. Could this be because I wrapped the text after pasting it? Does it work without commas if all on one line?

Comment (hemidactylus) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:94]: I tried that as well and, to my puzzlement, it does not seem to work even in that case.

**Comment by alexellis** [comment]
Let's see if we can add 404 handling, and have it return an empty result?

https://github.com/openfaas/openfaas.github.io/pull/249#discussion_r708052722

File: _posts/2021-08-26-faas-storage-cassandra-astra.md (hunk 2 above)
Comment (hemidactylus) [reply _posts/2021-08-26-faas-storage-cassandra-astra.md:437]: In the Document API there is a specific endpoint for creating a collection, which goes something like this:
```
      curl -X POST \
        "https://<DB_ID>-<DB_REGION>.apps.astra.datastax.com/api/rest/v2/namespaces/functions/collections" \
        -H  "accept: application/json" \
        -H  "X-Cassandra-Token: <APP_TOKEN>" \
        -H  "Content-Type: application/json" \
        -d "{\"name\":\"COLLECTION_NAME\"}"
```
I am not sure this is exposed through the @astrajs/collections SDK, though (I can inquiry on this).

Adding the command above (to create "links" collection) to the setup instructions would make the functions work in all cases (i.e. even on a fresh install with no documents yet in the collection).

As said above, I'll look into the Node SDK about collection creation (it does not seem to be there, anyway).
